From: Mohan <mohanv@ti.com>
Date: Fri, 6 May 2011 10:50:05 +0000 (+0530)
Subject: omap: serial: save and restore LCR
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=43df1359a7a67f3834ecd9733569f6ef806257f2

omap: serial: save and restore LCR

Currently the CHAR_LENGTH of LCR is hardcoded to 8-bits
while restoring uart registers. Because of this, if
CHAR_LENGTH is changed to other than 8, garbage characters
are seen on the console when the system enters and exits idle mode.

Saving and restoring the LCR register fixes the issue.

Change-Id: I4502735b97d9ff7fc87d3c11a9bd8497894d4ab0
Signed-off-by: Mohan V <mohanv@ti.com>
Signed-off-by: Maulik Mankad <maulik@ti.com>
---

Index: kernel/arch/arm/mach-omap2/serial.c
===================================================================
--- kernel.orig/arch/arm/mach-omap2/serial.c	2011-05-21 00:43:57.309639169 +0900
+++ kernel/arch/arm/mach-omap2/serial.c	2011-05-21 00:44:49.220846910 +0900
@@ -114,6 +114,7 @@
 	u16 mcr;
 	u16 mdr3;
 	u16 dma_thresh;
+	u16 lcr;
 
 
 #endif
@@ -249,6 +250,7 @@
 		return;
 
 	lcr = serial_read_reg(uart, UART_LCR);
+	uart->lcr = lcr;
 	serial_write_reg(uart, UART_LCR, OMAP_UART_LCR_CONF_MDB);
 	uart->dll = serial_read_reg(uart, UART_DLL);
 	uart->dlh = serial_read_reg(uart, UART_DLM);
@@ -313,7 +315,7 @@
 	/* Config B mode */
 	serial_write_reg(uart, UART_LCR, OMAP_UART_LCR_CONF_MDB);
 	serial_write_reg(uart, UART_EFR, efr);
-	serial_write_reg(uart, UART_LCR, UART_LCR_WLEN8);
+	serial_write_reg(uart, UART_LCR, uart->lcr);
 	serial_write_reg(uart, UART_OMAP_SCR, uart->scr);
 	serial_write_reg(uart, UART_OMAP_WER, uart->wer);
 	serial_write_reg(uart, UART_OMAP_SYSC, uart->sysc);
